### PR TITLE
Implement Promise.allSettled

### PR DESCRIPTION
**Pull request by boingoing**
Implement Promise.allSettled

This is now in Stage 3 and supported in V8 and JSC. See spec:
https://tc39.github.io/proposal-promise-allSettled/

Fixes: #6056

### CODE REVIEW

File: lib/Runtime/Library/JavascriptPromise.cpp [hunk 1]
```diff
@@ -274,6 +274,148 @@ namespace Js
...
+        }
+
+        remainingElementsWrapper->remainingElements--;
+
+        // We want this call to happen outside the try statement because if it throws, we aren't supposed to reject the promise.
+        if (remainingElementsWrapper->remainingElements == 0)
+        {
+            Assert(values != nullptr);
+
+            TryCallResolveOrRejectHandler(promiseCapability->GetResolve(), values, scriptContext);
```
[the remark this answers is not in the document]
Comment (boingoing) [reply lib/Runtime/Library/JavascriptPromise.cpp]: I think you are right, good catch. We have the same issue with `Promise.all`, it looks like. I'll make the fix to both.

**Comment by fatcerberus** [comment]
> it seems that if this call throws, then it is "caught" by allSettled step 7, which then rejects the resulting promise with the exception value.

Wait, is that actually right?  I thought the whole deal with `Promise.allSettled` was that it never rejects, so that you can see the results of all promises regardless of fulfill/reject unlike `Promise.all`?

**Comment by boingoing** [comment]
@zenparsing I don't actually know how to retrigger the tests which failed in the new CI. These failures aren't related to my change, though, they've been failing across the CI lately. I'll rebase the branch which will trigger them to rerun, at least.

**Comment by boingoing** [comment]
@fatcerberus Yes, this reading of the spec is correct. We call `IfAbruptRejectPromise` mostly on errors thrown out of the iterator or on errors invoking the `resolve` method. The case @zenparsing caught is what happens when the iterable argument is empty and an error is thrown while trying to execute `resolve` to get a promise resolved with the empty array.

**Pull request merged**